### PR TITLE
Single factory interfaces

### PR DESCRIPTION
**Pull request by shadowhand**
As per https://github.com/php-fig/fig-standards/pull/759#issuecomment-238969730

### CODE REVIEW

**Comment by shadowhand** [comment]
@mindplay-dk does this satisfy your concerns?

File: src/Stream/ResourceStreamFactoryInterface.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+namespace Psr\Http\Factory\Stream;
+
+use Psr\Http\Message\StreamInterface;
+
+interface StreamFactoryInterface
```
Comment (nyamsprod) [inline src/Stream/ResourceStreamFactoryInterface.php]: the interface name should be `ResourceStreamFactoryInterface`

File: src/Stream/StringStreamFactoryInterface.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+namespace Psr\Http\Factory\Stream;
+
+use Psr\Http\Message\StreamInterface;
+
+interface StreamFactoryInterface
```
Comment (nyamsprod) [inline src/Stream/StringStreamFactoryInterface.php]: The interface name should be `StringStreamFactoryInterface`

File: src/Stream/ResourceStreamFactoryInterface.php (hunk 1 above)
Comment (shadowhand) [inline src/Stream/ResourceStreamFactoryInterface.php]: Fixed.

File: src/Stream/StringStreamFactoryInterface.php (hunk 2 above)
Comment (shadowhand) [inline src/Stream/StringStreamFactoryInterface.php]: Also fixed, thanks for catching that!

File: src/Stream/StreamFactoryInterface.php [hunk 3]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+namespace Psr\Http\Factory\Stream;
+
+use Psr\Http\Message\StreamInterface;
+
+interface StreamFactoryInterface
```
Comment (mindplay-dk) [inline src/Stream/StreamFactoryInterface.php:7]: I still think this interface is duplicating `StringStreamFactoryInterface` - both create a string-based stream, so the implementations are going to be identical, only the initialization differs, and you can cover that by adding a default value of `''` to the argument.

File: src/Request/RequestFactoryInterface.php [hunk 4]
```diff
@@ -1,6 +1,6 @@
 <?php
 
-namespace Psr\Http\Factory;
+namespace Psr\Http\Factory\Request;
```
Comment (mindplay-dk) [inline src/Request/RequestFactoryInterface.php:3]: Do we really need all this name littering? All these namespaces with just one or two interfaces. I think keeping all factories in a single `Factory` sub-namespace was just fine.

Comment (shadowhand) [inline src/Request/RequestFactoryInterface.php:3]: When there was only one of each interface it seemed fine, but then separating them made it into a mess:

```
CallbackStreamFactoryInterface.php
GlobalServerRequestFactoryInterface.php
RequestFactoryInterface.php
ResourceStreamFactoryInterface.php
ResponseFactoryInterface.php
ServerRequestFactoryInterface.php
StreamFactoryInterface.php
StringStreamFactoryInterface.php
UploadedFileFactoryInterface.php
UriFactoryInterface.php
```

Streams are mixed in all over the place and I don't really like `StreamFactoryCallbackInterface` so having a separate directory for each seemed to make more sense.

Comment (shadowhand) [inline src/Request/RequestFactoryInterface.php:3]: With a separate directory for each type:

```
Request/RequestFactoryInterface.php
Response/ResponseFactoryInterface.php
ServerRequest/GlobalServerRequestFactoryInterface.php
ServerRequest/ServerRequestFactoryInterface.php
Stream/CallbackStreamFactoryInterface.php
Stream/ResourceStreamFactoryInterface.php
Stream/StreamFactoryInterface.php
Stream/StringStreamFactoryInterface.php
UploadedFile/UploadedFileFactoryInterface.php
Uri/UriFactoryInterface.php
```

File: src/Stream/StreamFactoryInterface.php (hunk 3 above)
Comment (shadowhand) [inline src/Stream/StreamFactoryInterface.php]: That's a good point, I'll update.

Comment (shadowhand) [inline src/Stream/StreamFactoryInterface.php:7]: Makes sense to me. There's no way to attach a stream after `StreamInterface` is created, so creating an empty stream is fairly pointless and `ResourceStreamFactoryInterface` covers that case.

File: src/Stream/StreamFactoryInterface.php [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+     * Create a new stream from a string.
+     *
+     * A temporary resource will be created with the content of the string.
+     * The resource will be writable and seekable.
+     *
+     * @param string $body
+     *
+     * @return StreamInterface
+     */
+    public function createStream($body = '');
```
Comment (shadowhand) [inline src/Stream/StreamFactoryInterface.php:19]: @mindplay-dk updated.

**Comment by nyamsprod** [comment]
I have a question regarding this PR. why can't we have a different approach. Because right now it seems like we will end up with a list of infinite discrete StreamFactory interfaces. Why can't we just "copy" the IteratorAggregate interface.

``` php

namespace Psr\Http\Factory;

use Psr\Http\Message\StreamInterface;

interface StreamFactoryInterface
{
    /**
     * Create a new stream
     *
     * @return StreamInterface
     */
    public function createStream();

}
```

And then use a util repo like the ones used by [PSR-6](https://github.com/php-fig/cache-util) and  [PSR-7](https://github.com/php-fig/http-message-util) to extends this interface with more specific interfaces of each type of stream.

This way we don't clutter the PSR with a unfinished list of stream factory and developer can either typecast against the general `StreamFactoryInterface` if they don't care about the stream returned or against a more restrictive interface found in the Util package which is easier to expand.

**Comment by shadowhand** [comment]
@mindplay-dk what do you think of the approach suggested by @nyamsprod? Seems like it might be a reasonable approach, at least with streams. I do think having the global server request interface is still useful though.

File: src/Request/RequestFactoryInterface.php (hunk 4 above)
Comment (mindplay-dk) [inline src/Request/RequestFactoryInterface.php:3]: I don't see the need to "group" single classes, those extra namespaces are redundant. I would group together the Request factories, and leave the URI and uploaded file factories in the root namespace.

```
Request/RequestFactoryInterface.php
Request/ServerRequestFactoryInterface.php
Request/GlobalServerRequestFactoryInterface.php
Response/ResponseFactoryInterface.php
Stream/CallbackStreamFactoryInterface.php
Stream/ResourceStreamFactoryInterface.php
Stream/StreamFactoryInterface.php
Stream/StringStreamFactoryInterface.php
UploadedFileFactoryInterface.php
UriFactoryInterface.php
```

Or even just group the streams and leave everything else.

```
RequestFactoryInterface.php
ServerRequestFactoryInterface.php
GlobalServerRequestFactoryInterface.php
ResponseFactoryInterface.php
UploadedFileFactoryInterface.php
UriFactoryInterface.php
Stream/CallbackStreamFactoryInterface.php
Stream/ResourceStreamFactoryInterface.php
Stream/StreamFactoryInterface.php
Stream/StringStreamFactoryInterface.php
```

Dedicated namespaces for individual factories seems a little overkill.

**Comment by mindplay-dk** [comment]
> what do you think of the approach suggested by @nyamsprod?

Read it over and over, I don't understand what he's suggesting, sorry.

**Comment by shadowhand** [comment]
@mindplay-dk as I understand it, @nyamsprod is suggesting that we drop the extra stream factory interfaces from the main specification and create a separate package for them, such as `http-factory-streams`.

File: src/Request/RequestFactoryInterface.php (hunk 4 above)
Comment (harikt) [inline src/Request/RequestFactoryInterface.php:3]: I like the grouping for it is easy to look less files. Eg : When `Request` and `Response` are on same directory it is hard. Hard time to read the 2nd one. @mindplay-dk  the first one seems much better in that case.

**Comment by shadowhand** [comment]
I'm going to close this and take a different approach.

**Comment by mindplay-dk** [comment]
> as I understand it, @nyamsprod is suggesting that we drop the extra stream factory interfaces from the main specification and create a separate package for them

@shadowhand that seems really overkill - we're talking about a few basic stream creation methods, requiring several install just for interfaces seems silly. Adding a new factory interface is not a backwards breaking change, and with 90% certainty, the ones implemented by e.g. Diactoros covers everything anybody will ever need.

**Pull request closed without merging**